packages/dd-trace/src/collector/encoder.js

```diff
@@ -0,0 +1,585 @@
+  }
+
+  encodeSegmentDiscard (event) {
+    const bytes = this._eventBytes
+
+    this._beforeEncode(eventTypes.START_SEGMENT, event)
+
+    this._encodeUnsigned(bytes, eventTypes.DISCARD_SEGMENT)
+    this._encodeFixArray(bytes, 1)
+    this._encodeSegmentId(bytes, event.segmentId)
+
+    this._afterEncode()
+  }
+
+  encodeSpanStart (event) {
+    const bytes = this._eventBytes
+
+    this._beforeEncode(eventTypes.START_SPAN, event)
+
+    this._encodeUnsigned(bytes, eventTypes.START_SPAN)
+    this._encodeFixArray(bytes, 10)
+    this._encodeTime(bytes, event.ticks)
+    this._encodeSegmentId(bytes, event.segmentId)
+    this._encodeId(bytes, event.spanId)
+    this._encodeUnsigned(bytes, event.parentIndex)
+    this._encodeString(bytes, event.service)
+    this._encodeString(bytes, event.name)
+    this._encodeString(bytes, event.resource)
+    this._encodeMeta(bytes, event.meta)
+    this._encodeMetrics(bytes, event.metrics)
+    this._encodeString(bytes, event.type)
+
+    this._afterEncode()
+  }
+
+  encodeSpanFinish (event) {
+    const bytes = this._eventBytes
+
+    this._beforeEncode(eventTypes.FINISH_SPAN, event)
+
+    this._encodeUnsigned(bytes, eventTypes.FINISH_SPAN)
+    this._encodeFixArray(bytes, 3)
+    this._encodeTime(bytes, event.ticks)
+    this._encodeSegmentId(bytes, event.segmentId)
+    this._encodeUnsigned(bytes, event.spanIndex)
+
+    this._afterEncode()
+  }
+
+  encodeAddTags (event) {
+    const bytes = this._eventBytes
+
+    this._beforeEncode(eventTypes.ADD_TAGS, event)
+
+    this._encodeUnsigned(bytes, eventTypes.ADD_TAGS)
+    this._encodeFixArray(bytes, 4)
+    this._encodeSegmentId(bytes, event.segmentId)
+    this._encodeUnsigned(bytes, event.spanIndex)
+    this._encodeMeta(bytes, event.meta)
+    this._encodeMetrics(bytes, event.metrics)
+
+    this._afterEncode()
+  }
+
+  encodeException (event) {
+    const bytes = this._eventBytes
+
+    this._beforeEncode(eventTypes.EXCEPTION, event)
+
+    this._encodeUnsigned(bytes, eventTypes.EXCEPTION)
+    this._encodeFixArray(bytes, 5)
+    this._encodeSegmentId(bytes, event.segmentId)
+    this._encodeUnsigned(bytes, event.spanIndex)
+    this._encodeString(bytes, event.error.message || '')
+    this._encodeString(bytes, event.error.name || '')
+    this._encodeString(bytes, event.error.stack || '')
+
+    this._afterEncode()
+  }
+
+  makePayload () {
+    const stringSize = this._stringBytes.length + 6
+    const segmentSize = this._segmentBytes.length + 6
+    const eventSize = this._eventBytes.length
+    const buffer = Buffer.allocUnsafe(stringSize + segmentSize + eventSize)
+
+    let offset = 0
+
+    offset = this._writeStrings(buffer, offset)
+    offset = this._writeSegments(buffer, offset)
+    this._writeEvents(buffer, offset)
+
+    this._reset()
+
+    return buffer
+  }
+
+  flush (done = () => {}) {
+    try {
+      const data = this.makePayload()
+
+      this._timer = clearTimeout(this._timer)
+
+      collector.send_events(data)
```

Comment:
❓ is `collector.send_events` sync? Do we need a `done` callback then? Can't we just return? Or is this done to preserve the `flush` signature from other exporters?